### PR TITLE
golang filter extension: add recover in defer to catch panic in go code

### PR DESCRIPTION
**Pull request by doujiang24**
Commit Message: 
Additional Description:
fix https://github.com/envoyproxy/envoy/issues/25288

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by wangfakang — COMMENTED** [review]
Thanks. good works.

**Comment by wangfakang** [comment]
> contrib only update. ping me when it's readly to merge cc @wangfakang

OK. thanks.

File: contrib/golang/filters/http/source/go/pkg/http/shim.go [hunk 1]
```diff
@@ -115,6 +115,13 @@ func envoyGoFilterOnHttpHeader(r *C.httpRequest, endStream, headerNum, headerByt
 			req = createRequest(r)
 		}
 	}
+	if req.paniced {
+		req.safeReplyPanic()
+		// only may hit it when filter is just destroyed
+		// status is meaningless then.
+		return uint64(api.Continue)
+	}
+	defer req.RecoverPanic()
```
[the remark this answers is not in the document]
Comment (doujiang24) [reply contrib/golang/filters/http/source/go/pkg/http/shim.go:124]: this is the first place that we have `req`, I don't think we could move it forward.
Also, the code above is belong the Envoy Go extension, we could assume they won't panic, which is not the `RecoverPanic ` want to protect.

File: contrib/golang/filters/http/source/go/pkg/http/filter.go [hunk 2]
```diff
@@ -41,6 +41,35 @@ import (
...
+			// will delay to the next time entering Go.
+			r.paniced = true
+
+		default:
+			// The following safeReplyPanic should only may get ErrRequestFinished,
+			// ErrFilterDestroyed or ErrNotInGo, won't hit this branch, so, won't dead loop here.
+
+			// ErrInvalidPhase, or panic from other places, not from not-ok C return status.
+			// It's safe to try send a local reply with 500 status.
+			r.safeReplyPanic()
```
[the remark this answers is not in the document]
Comment (doujiang24) [reply contrib/golang/filters/http/source/go/pkg/http/filter.go:70]: Yep, done.

File: contrib/golang/filters/http/source/go/pkg/api/filter.go [hunk 3]
```diff
@@ -70,6 +70,8 @@ type FilterCallbacks interface {
 	// Continue or SendLocalReply should be last API invoked, no more code after them.
 	Continue(StatusType)
 	SendLocalReply(responseCode int, bodyText string, headers map[string]string, grpcStatus int64, details string)
+	// RecoverPanic recover panic in defer
+	RecoverPanic()
```
[the remark this answers is not in the document]
Comment (doujiang24) [reply contrib/golang/filters/http/source/go/pkg/api/filter.go:74]: This is the low-level API, we could add another user-friendly wrapper in the feature, at a higher level, i.e. the async default level API wrapper, as discussed offline.

**Review by wangfakang — APPROVED** [review]
LGTM. ping @wbpcode

**Review by wbpcode — APPROVED** [review]
Thanks.

**Pull request merged**